# scripts_for_my_daily_work

## [crabz](https://github.com/sstadick/crabz)

```shell
# Compress
tar -I 'crabz -l 9 -p 12' -cvf <archive>.tar.gz <input_dir>
tar -cf - <input_dir> | crabz -l 9 -p 12 > <archive>.tar.gz
# Decompress
crabz -d -p 12 <archive>.tar.gz | tar -xvf -
```

## [pigz](https://github.com/madler/pigz)

```shell
# Compress
tar -I 'pigz -9 -p 12' -cvf <archive>.tar.gz <input_dir>
tar -cvf - <input_dir> | pigz -9 -p 12 > <archive>.tar.gz
# Decompress
tar -I 'pigz -d -p 12' -xvf <archive>.tar.gz
pigz -d -p 12 <archive>.tar.gz && tar -xvf <archive>.tar && rm <archive>.tar
```

## [AnnoSINE_v2](https://github.com/liaoherui/AnnoSINE_v2)

```shell
mamba activate AnnoSINE
# If set arg "-a" to 0, then Hmmer will search SINE using the plant hmm file. (default: 0)
python $ANNOSINE_HOME/bin/AnnoSINE_v2.py -t 8 -a 1 -f 'y' 3 <input_filename> <output_dir>
script -c 'python $ANNOSINE_HOME/bin/AnnoSINE_v2.py -t 8 -a 1 -f y 3 <input_filename> <output_dir>' <output_dir>/<output_filename>.script.log

```

## extract_one_gene_from_a_gff_file.sh

ä»ä¸€ä¸ªGFF3æ–‡ä»¶ä¸­æå–æŒ‡å‘æŸä¸ªåŸºå› çš„æ•°è¡Œã€‚å®åˆ—ï¼š

```shell
for i in $(<{input.id_list}); do echo "$i"; ./extract_one_gene_from_a_gff_file.sh -q "$i" -s {input.ori_gff} >> {output}; done
# å¦ä¸€ä¸ªåŠæ³•
while read line; do id=$(echo "$line" | sd '[prefix or suffix]' ''); echo "$id"; rg "$id" ./[input.gff] | rg '\tgene' > ./[input.gene.gff]; done < [input.id_list]
```

## cal_fasta_seq_len.py

ç»Ÿè®¡ä¸€ä¸ªFASTAæ–‡ä»¶ä¸­æ‰€æœ‰åºåˆ—çš„é•¿åº¦ç„¶åä»¥CSVæ ¼å¼è¾“å‡ºï¼Œä¸€èˆ¬é€‰æœ€é•¿çš„ä½œä¸ºå‚è€ƒåºåˆ—ã€‚ç¤ºä¾‹ï¼š

```shell
python3 ./cal_fasta_seq_len.py -i ./XXX_protein.faa -o ./XXX_pro_seq_len.csv
grep '>' ./XXX_protein.faa
# è¿›è¡Œå¤šåºåˆ—æ¯”å¯¹
clustalo -i ./XXX_protein.faa -o ./XXX_pro_MSA_result.clustal --outfmt=clustal --full --iterations=5 -v --force
```

å°† .csv ä¸ .clustal ä¸¤ä¸ªæ–‡ä»¶ç»“åˆåˆ¤æ–­ï¼Œå¯ä»¥é€‰å‡ºåˆé€‚çš„åºåˆ—ç”¨äºè¿›ä¸€æ­¥ç ”ç©¶ã€‚

## vina_docking-log_parser.py

å¤„ç†AutoDock Vinaçš„å¯¹æ¥æ—¥å¿—ã€‚ä¼ å…¥å¯¹æ¥ç»“æœï¼ˆè¾“å‡ºçš„logï¼‰ï¼Œæ­¤è„šæœ¬çš„è¾“å‡ºæ–‡ä»¶ä¸ºä¸€ä¸ªCSVæ–‡ä»¶ï¼Œå®ƒä»¶åŒ…å«æœ‰å…³å¯¹æ¥modeçš„äº²å’ŒåŠ›ã€RMSDä¿¡æ¯ã€‚

```
mode |   affinity | dist from best mode
     | (kcal/mol) | rmsd l.b.| rmsd u.b.
-----+------------+----------+----------
   1       -4.091          0          0
   2        -3.98      2.247       2.37
   3       -3.915       16.1       16.9
   ...
```

ğŸ‘‡

| mode | affinity (kcal/mol) | dist from best mode (rmsd l.b.) | dist from best mode (rmsd u.b.) |
|------|---------------------|---------------------------------|---------------------------------|
| 1    | -4.091              | 0                               | 0                               |
| 2    | -3.98               | 2.247                           | 2.37                            |
| 3    | -3.915              | 16.1                            | 16.9                            |
| ...  | ...                 | ...                             | ...                             |

ç¤ºä¾‹ï¼š

```shell
python3 ./vina_docking-log_parser.py -i ./XXX.vina.out.log -o ./XXX.vina.out.log.csv
```

ç„¶åæŸ¥çœ‹CSVæ–‡ä»¶å°±å¯ä»¥çŸ¥é“äº²å’ŒåŠ›ã€RMSDäº†â˜ºï¸

## æˆ‘çš„ Docker å¸¸ç”¨æŒ‡ä»¤

```
docker container ls  # æŸ¥çœ‹éœ€è¦pushçš„container
docker commit -a "Yuhao Cai" e1079b1cf715 caicai0/my_debian:v1.2  # â€œe1079b1cf715â€æ˜¯containerçš„IDã€‚
docker push caicai0/my_debian:v1.2
# æ¥ä¸‹æ¥æ˜¯å¯åŠ¨pullä¸‹æ¥çš„å®¹å™¨
docker run -it -d --name="nvidia-devel_ubuntu22.04" --hostname="caicai" --volume="C:\Users\cauca\Desktop\work":"/home/work" --workdir="/home/work" --privileged=true --gpus all nvidia/cuda:12.3.1-devel-ubuntu22.04
docker run -it -d --name='nvidia-devel_ubuntu22.04' --hostname='caicai' --volume='C:\Users\tx4cai\Desktop\work':'/home/work' --workdir='/home/work' --privileged=true --gpus all -e 'TZ=Asia/Shanghai' --publish-all 'caicai0/my_bio_env-nvidia-cuda-12.3.1-devel-ubuntu22.04:v1.5_for_2024-06-03'

# æ²¡å¿…è¦çš„å¦‚ä¸‹ï¼š
docker run -it -d --name="nvidia-devel_ubuntu22.04" --hostname="caicai" --volume="C:\Users\cauca\Desktop\work":"/home/work" --workdir="/home/work" --privileged=true --gpus all -e NVIDIA_DRIVER_CAPABILITIES=compute,utility -e NVIDIA_VISIBLE_DEVICES=all nvidia/cuda:12.3.1-devel-ubuntu22.04
```

- å¼ºçƒˆå»ºè®®ä½¿ç”¨ GPUï¼Œ[NVIDIA Dockerï¼šè®© GPU æœåŠ¡å™¨åº”ç”¨ç¨‹åºéƒ¨ç½²å˜å¾—å®¹æ˜“](https://developer.nvidia.com/zh-cn/blog/nvidia-docker-gpu-server-application-deployment-made-easy/)

## ä½¿ç”¨ Docker-JupyterLab-R

```
docker run -it -d --name="my_bio_env_jupyter" --hostname="caicai" --user root --volume="/home/caicai/work":"/home/work" --workdir="/home/work" --privileged=true -e TZ=Asia/Shanghai --publish-all docker.io/jupyter/r-notebook:x86_64-r-4.3.1 /bin/bash -c "jupyter lab --allow-root"
```

ç„¶ååœ¨ Podman Desktop æŸ¥çœ‹å¯åŠ¨çš„å®¹å™¨ï¼Œåœ¨ `Containers > Container Details` çš„å³ä¸Šè§’ç‚¹å‡»â€œâ†—â€ï¼ˆOpen Browserï¼‰ï¼Œå†æ‰§è¡Œ `jupyter server list` å‘½ä»¤è·å– token ä»¥ç™»å½•ï¼Œå¦‚ä¸‹å›¾ï¼š

![è¿™æ˜¯å›¾ç‰‡](/images/podman_for_R.jpg "podman for R")
![è¿™æ˜¯å›¾ç‰‡](/images/JupyterLab_for_R.jpg "JupyterLab for R")

## MSA è½¯ä»¶å¸¸ç”¨å‘½ä»¤

### [PROBCONS](http://probcons.stanford.edu/)

- ä½¿ç”¨mambaå®‰è£…ã€‚

### [MAFFT](https://gitlab.com/sysimm/mafft)

> [ä½¿ç”¨ MAFFT è¿›è¡Œå¤šåºåˆ—æ¯”å¯¹](http://www.chenlianfu.com/?p=2214)
> [](https://mafft.cbrc.jp/alignment/software/manual/manual.html#lbAI)
- ç¼–è¯‘å®Œæˆåéœ€è¦å…ˆæ‰§è¡Œè½¯é“¾æ¥å‘½ä»¤ï¼Œä»¥ä½¿ç”¨ `mafft --man`ï¼š

```shell
$ mafft --man
man: /root/mafft-master/libexec/mafft/mafft.1: No such file or directory
No manual entry for /root/mafft-master/libexec/mafft/mafft.1
$ fd "mafft.1" ~/mafft-master
/root/mafft-master/core/mafft.1
/root/mafft-master/share/man/man1/mafft.1
$ md5sum /root/mafft-master/core/mafft.1 /root/mafft-master/share/man/man1/mafft.1
1eadd020af7e4f50baaf479f70823408  /root/mafft-master/core/mafft.1
1eadd020af7e4f50baaf479f70823408  /root/mafft-master/share/man/man1/mafft.1
$ ln -s /root/mafft-master/share/man/man1/mafft.1 /root/mafft-master/libexec/mafft/mafft.1
$ mafft --man | head -n 5
MAFFT(1)                                   Mafft Manual                                  MAFFT(1)

THIS MANUAL IS FOR V6.2XX (2007)
       Recent versions (v7.1xx; 2013 Jan.) have more features than those described below.  See
       also the tips page at http://mafft.cbrc.jp/alignment/software/tips0.html
```

```shell
mafft --clustalout --auto [input.fasta] > [output.clustal]
script -c 'time mafft --maxiterate 1000 --auto [input.fasta] > [output.fasta]' [output.fasta.log]
```

### [T-Coffee](https://tcoffee.org/Projects/tcoffee/index.html)

ä¼šè¾“å‡º[ä¸‰ä¸ªæ–‡ä»¶](https://tcoffee.org/Projects/tcoffee/documentation/index.html#t-coffee)ã€‚

```
When aligning, T-Coffee will always at least generate three files:

        sample_seq1.aln : Multiple Sequence Alignment (ClustalW format by default)

        sample_seq1.dnd : guide tree (Newick format)

        sample_seq1.html : colored MSA according to consistency (html format)
```

```shell
t_coffee [input.fasta] -mode quickaln
```

### [Clustal Omega](http://www.clustal.org/omega/)

## [trimAl](https://vicfero.github.io/trimal/)

> [Welcome to trimAlâ€™s documentation!](https://trimal.readthedocs.io/en/latest/index.html)

```shell
trimal -in [input.fasta] -out[output.fasta] -automated1
```

## [IQ-TREE](https://github.com/iqtree/iqtree2) å¸¸ç”¨å‘½ä»¤

> [IQ-tree2æ„å»ºç³»ç»Ÿå‘è‚²æ ‘ä½¿ç”¨å°ç»“](https://www.jianshu.com/p/aa0192f37657)
å»ºè®®å•ç‹¬åˆ›ç«‹ä¸€ä¸ªè·¯å¾„ç”¨äºå­˜æ”¾å»ºæ ‘ç”¨çš„ \[input.fasta\]ã€\[MSAoutput.phy\]ã€[qitree.output ...]ã€‚

```shell
# æ— æ ¹æ ‘
iqtree2 -s [MSA-aligned-output.ply] -B 1000 -nt 16 --prefix my_iqtree2 -gz --redo-tree
script -c 'iqtree2-mpi -s [MSA-aligned-output.trimal.fa] -B 1000 -nt 12 --prefix my_iqtree2-mpi -gz --redo-tree' my_iqtree2-mpi.script.log
```

## PyMOL æ•™è‚²ç‰ˆï¼Œç”³è¯·èµ·æ¥æ¯”è¾ƒå®¹æ˜“

- [PyMOLæ•™è‚²ç‰ˆå®‰è£…æ•™ç¨‹](https://zhuanlan.zhihu.com/p/598711018)
- [PyMOL å®˜ç½‘](https://pymol.org/2/)ï¼Œä¸‹è½½ã€‚
- [PyMod å¦‚è™æ·»ç¿¼](https://pymolwiki.org/index.php/PyMod)ï¼Œå¯è§†åŒ–åœ°ä½¿ç”¨ modellerã€‚
- [èƒ½ç”¨ conda å®‰è£… modeller](https://salilab.org/modeller/download_installation.html) çš„è¯ï¼Œç»ä¸ç”¨è½¯ä»¶åº“æˆ–è§£å‹åŒ…ï¼Œå› ä¸ºç¯å¢ƒä¾èµ–é—®é¢˜ã€‚
- é€šè¿‡ conda åˆ›å»º pymol-open-source çš„è¿è¡Œç¯å¢ƒï¼Œå®‰è£… pymol-open-sourceã€biopythonï¼›ç„¶åå°† modeller å®‰è£…åœ¨ pymol-open-source çš„è¿™ä¸ªç¯å¢ƒã€‚
- å¯åŠ¨ pymol-open-source åï¼Œå®‰è£… pymod3 æ’ä»¶ï¼Œå¹¶é‡å¯ã€‚
- å¯åŠ¨ pymodï¼Œåœ¨å…¶å¯¼èˆªæ çš„ help é€‰é¡¹ä¸­ç‚¹å‡»â€Install PyMod Componentsâ€œï¼Œå®Œæˆå…¶ä½™å®‰è£…ã€‚
- ç„¶åå°±å¯ä»¥å€ŸåŠ© PyMod ä½¿ç”¨ Modeller è¿›è¡ŒåŒæºå»ºæ¨¡äº†ã€‚

## [imgcat](https://github.com/eddieantonio/imgcat) æŸ¥çœ‹å›¾åƒ

```shell
imgcat -d 24bit -RH my_github_avatar.jpg
```

å›¾åƒæ¸…æ™°åº¦æœ€å¥½ã€‚


## [termimage](https://github.com/nabijaczleweli/termimage) æŸ¥çœ‹å›¾åƒ

```shell
termimage -s [width]x[height] my_github_avatar.jpg
```
å›¾åƒæ¸…æ™°åº¦æ¬¡äº imgcatã€‚

## [viu](https://github.com/atanunq/viu) æŸ¥çœ‹å›¾åƒ

```shell
viu -b my_github_avatar.jpg
```

å›¾åƒç€è‰²ååˆ†â€œå¤å¤â€ã€‚

## ä¸€äº› rust å‘½ä»¤è¡Œå·¥å…·

```
bat    btm    cargo-binstall  cargo-fmt   clippy-driver  difft   dust  exa  lfs  mcfly  procs  rls            rust-gdb     rust-lldb  rustdoc  rustup     tldr      viu  zoxide
broot  cargo  cargo-clippy    cargo-miri  delta          dprint  eva   fd   lsd  ouch   rip    rust-analyzer  rust-gdbgui  rustc      rustfmt  termimage  topgrade  xcp 
```

---

## æˆ‘æƒ³ç© [VIM Adventures](https://zhuanlan.zhihu.com/p/628613725)

### ç ´è§£ç‰ˆ

```shell
git clone https://github.com/AkshayGupta8/Vim-adventure.git
cd Vim-adventure/vim-adventures
conda create -n vim-adventure -y
conda activate vim-adventure
conda install -c conda-forge nodejs -y
npm install -g npm@6
npm install
node bin/www.js
cd ..
cd ..
conda deactivate
conda env remove --name vim-adventure
rm -rf Vim-adventure
```

### [Interactive Vim tutorial](https://www.openvim.com/)ï¼Œå¥½ï¼

### [Terminus](http://web.mit.edu/mprat/Public/web/Terminus/Web/main.html)

## æ‰¹é‡ä¸‹è½½hmm

```shell
hmm_path='./hmm_files'

while read line; do
  id=$(echo $line | tr -d '\n')
  printf "Downloading %s ... " $id
  curl "https://www.ebi.ac.uk/interpro/wwwapi//entry/pfam/$($id)\?annotation\=hmm" -o "${hmm_path}/${line}".hmm.gz
  echo 'Done.'
  done < ./PFXXXXX.IDlist.txt

```

## [seqkit](https://bioinf.shenwei.me/seqkit/)

### ä½¿ç”¨`seqkit replace`é‡å‘½åFASTA ID

```
$ cat test.fa
>seq1 desc1
CCCCAAAACCCCATGATCATGGATC
>seq2 desc2
CCCCAAAACCCCATGGCATCATTCA
>seq3 desc3
CCCCAAAACCCCATGTTGCTACTAG

$ cat test.0.fa
>seq1
CCCCAAAACCCCATGATCATGGATC
>seq2
CCCCAAAACCCCATGGCATCATTCA
>seq3
CCCCAAAACCCCATGTTGCTACTAG

$ cat renameID.tsv
seq1     renamed_seq1
seq2     renamed_seq2
seq3     renamed_seq3

# æ ¹æ®kv-fileæ˜ å°„æ–‡ä»¶ï¼Œåªä¿®æ”¹IDï¼Œå¹¶ä¿ç•™æè¿°
$ seqkit replace -p '^(\S+) (.+?)$' -r '{kv} ${2}' -k renameID.tsv test.fa
[INFO] read key-value file: renameID.tsv
[INFO] 3 pairs of key-value loaded
>renamed_seq1 name1
CCCCAAAACCCCATGATCATGGATC
>renamed_seq2 name2
CCCCAAAACCCCATGGCATCATTCA
>renamed_seq3 name3
CCCCAAAACCCCATGTTGCTACTAG

# æ ¹æ®kv-fileæ˜ å°„æ–‡ä»¶ï¼Œåªä¿®æ”¹IDï¼Œå¹¶ä¿ç•™æ—§çš„IDå’Œæè¿°
$ seqkit replace -p '^(\S+) (.+?)$' -r '{kv} ${1} ${2}' -k renameID.tsv test.fa
[INFO] read key-value file: renameID.tsv
[INFO] 3 pairs of key-value loaded
>renamed_seq1 seq1 name1
CCCCAAAACCCCATGATCATGGATC
>renamed_seq2 seq2 name2
CCCCAAAACCCCATGGCATCATTCA
>renamed_seq3 seq3 name3
CCCCAAAACCCCATGTTGCTACTAG

# å¯¹äºç¼ºå°‘æè¿°éƒ¨åˆ†çš„FASTAï¼Œæ ¹æ®kv-fileæ˜ å°„æ–‡ä»¶ï¼Œåªä¿®æ”¹ID
$ seqkit replace -p '^(\S+)$' -r '{kv}' -k renameID.tsv test.0.fa
[INFO] read key-value file: renameID.tsv
[INFO] 3 pairs of key-value loaded
>renamed_seq1
CCCCAAAACCCCATGATCATGGATC
>renamed_seq2
CCCCAAAACCCCATGGCATCATTCA
>renamed_seq3
CCCCAAAACCCCATGTTGCTACTAG

# å¯¹äºç¼ºå°‘æè¿°éƒ¨åˆ†çš„FASTAï¼Œæ ¹æ®kv-fileæ˜ å°„æ–‡ä»¶ï¼Œåªä¿®æ”¹IDï¼Œå¹¶ä¿ç•™æ—§çš„ID
$ seqkit replace -p '^(\S+)$' -r '{kv} ${1}' -k renameID.tsv test.0.fa
[INFO] read key-value file: renameID.tsv
[INFO] 3 pairs of key-value loaded
>renamed_seq1 seq1
CCCCAAAACCCCATGATCATGGATC
>renamed_seq2 seq2
CCCCAAAACCCCATGGCATCATTCA
>renamed_seq3 seq3
CCCCAAAACCCCATGTTGCTACTAG

awk -F'\t' 'NR==FNR {map[$1]=$2; next} {if ($1 in map) print map[$1]; else print $1}' ids.map.tsv ids.txt > ids.renamed.txt
awk -F'\t' 'BEGIN{OFS="\t"} NR==FNR {map[$1]=$2; next} /^#/ {print; next} ($1 in map) {$1=map[$1]} {print}' id.map.tsv old.gff > new.gff

```

## [DESeq2](https://bioconductor.org/packages//release/bioc/html/DESeq2.html)

å®‰è£…è¿™ä¸ªRåŒ…ï¼š

```
mamba create -n DESeq2
mamamba activate DESeq2
mamba install -c bioconda bioconductor-genomeinfodb  # å…ˆè£…GenomeInfoDbï¼Œé¡ºåºä¸èƒ½å˜
mamba install -c bioconda -c conda-forge bioconductor-deseq2

#   bioconductor-biobase               2.66.0        r44h3df3fcb_0         bioconda
#   bioconductor-biocgenerics          0.52.0        r44hdfd78af_3         bioconda
#   bioconductor-biocparallel          1.40.0        r44he5774e6_1         bioconda
#   bioconductor-data-packages         20250625      hdfd78af_0            bioconda
#   bioconductor-delayedarray          0.32.0        r44h3df3fcb_1         bioconda
#   bioconductor-deseq2                1.46.0        r44he5774e6_1         bioconda
#   bioconductor-genomeinfodb          1.42.0        r44hdfd78af_2         bioconda
#   bioconductor-genomeinfodbdata      1.2.13        r44hdfd78af_0         bioconda
#   bioconductor-genomicranges         1.58.0        r44h3df3fcb_2         bioconda
#   bioconductor-iranges               2.40.0        r44h3df3fcb_2         bioconda
#   bioconductor-matrixgenerics        1.18.0        r44hdfd78af_0         bioconda
#   bioconductor-s4arrays              1.6.0         r44h3df3fcb_1         bioconda
#   bioconductor-s4vectors             0.44.0        r44h3df3fcb_2         bioconda
#   bioconductor-sparsearray           1.6.0         r44h3df3fcb_1         bioconda
#   bioconductor-summarizedexperiment  1.36.0        r44hdfd78af_0         bioconda
#   bioconductor-ucsc.utils            1.2.0         r44h9ee0642_1         bioconda
#   bioconductor-xvector               0.46.0        r44h15a9599_2         bioconda
#   bioconductor-zlibbioc              1.52.0        r44h3df3fcb_2         bioconda

R

library(DESeq2)

```

## [Minimap2](https://github.com/lh3/minimap2)

```
script -c 'minimap2 -t 12 -o minimap2_out.paf <target.fa> [query.fa]' minimap2_out.log
```

## [dotPlotly](https://github.com/tpoorten/dotPlotly)

```
git clone https://github.com/Rhylan2333/dotPlotly.git
script -c '../dotPlotly/pafCoordsDotPlotly.R -i minimap2_out.paf -o dotPlotly.minimap2.plot -k [number of autosomes and sex chromosomes + 1] -s -t -l -p 10' dotPlotly.minimap2.plot.log
```
